### PR TITLE
json: don't copy strings while escaping

### PR DESCRIPTION
**Pull request by mbolivar**
Re-implement json_escape() with a version that doesn't make a copy of the input string. Add test cases.

### CODE REVIEW

File: lib/json/json.c [hunk 1]
```diff
@@ -714,7 +673,38 @@ ssize_t json_escape(char *str, size_t *len, size_t buf_size)
...
+	/*
+	 * By walking backwards in the buffer from the end positions
+	 * of both the original and escaped strings, we avoid using
+	 * extra space. Characters in the original string are
+	 * overwritten only after they have already been escaped.
+	 */
+	str[escaped_len] = '\0';
+	next = &str[*len - 1];
+	dest = &str[escaped_len - 1];
+	while (1) {
```
Comment (lpereira) [inline lib/json/json.c]: Nitpick, but could be a `for (; next != str; next--)` instead.  Can remove the last few lines in the loop body.

**Review by lpereira — APPROVED** [review]
LGTM with a nitpick.

File: lib/json/json.c (hunk 1 above)
Comment (mbolivar) [inline lib/json/json.c]: I tried and failed to think of a way to do it without while(1) / break. The edge cases are a little tricky.

The way you suggest won't work if `*len == 1` and the character at `str[0]` needs escaping, for example.

**Comment by mbolivar** [comment]
Update:

- rewrote as a for loop
- added a test case to catch the edge condition discussed in last night's comments

**Review by lpereira — APPROVED** [review]

**Review by pfalcon — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**